### DISCUSSION TITLE
XML Parsing - Unusual XML Format

### DISCUSSION THREAD

**Discussion by vulnox** [grafana/grafana-infinity-datasource#316] [Q&A] [answered]
Hello,

I am trying to use Grafana/Infinity to display records from XML files on our network. This XML file is a bit unusual compared to what I see in other examples and I am having some issues getting it to pull the data. I was able to do the example XML (which was great) and get it to work, so the plugin seems to be set properly. I am not getting an CURL errors or anything to that effect, I can see the data from the XML in the Query Inspector.

So I assume then that the problem is either how I am trying to query the data out of the file, or our format is so unusual that it just may not work (I hope that isn't the case). The XML is used to display a proprietary "Health Monitor" that was created yeeeaaaars ago. The underlying health monitor itself still works well, but since it's basically an XML file with a style sheet, it only really works with Internet Explorer. As IE is finally going away, I am trying to use Grafana to display the same data from the existing XML format, since this HealthMon at this point is deployed to thousands of servers.

Thanks in advance!

Anyway, here is a snippet example of the XML:

```
<ItemsHeader>
	<Items>
		<Item itemType="Runtime Parameters" stateEntries="" lastChecked="5/19/2022 12:29:28 PM">
			<MonName>
				<Name misc="ANS" type="" state=""><![CDATA[AlertDestination]]></Name>
			</MonName>
			<MonName>
				<Name misc="14" type="" state=""><![CDATA[ArchiveExpiryDays]]></Name>
			</MonName>
			<MonName>
				<Name misc="LEFT" type="" state=""><![CDATA[CheckNameAppended]]></Name>
			</MonName>
			<MonName>
				<Name misc="False" type="" state=""><![CDATA[Debug]]></Name>
			</MonName>
		</Item>
		<Item itemType="Ping Results" stateEntries="" lastChecked="5/19/2022 12:29:28 PM">
			<MonName>
				<Name misc="OK" type="" state=""><![CDATA[Server1]]></Name>
			</MonName>
			<MonName>
				<Name misc="OK" type="" state=""><![CDATA[Server2]]></Name>
			</MonName>
			<MonName>
				<Name misc="OK" type="" state=""><![CDATA[Server3]]></Name>
			</MonName>
			<MonName>
				<Name misc="OK" type="" state=""><![CDATA[Server4]]></Name>
			</MonName>
		</Item>
		<Item itemType="Memory Usage" stateEntries="" lastChecked="5/19/2022 12:29:28 PM">
			<MonName>
				<Name misc="14.18MB | 14.18MB | 14.18MB" type="" state=""><![CDATA[FirstPAGE Alarm Manager Server]]></Name>
			</MonName>
			<MonName>
				<Name misc="44.21MB | 44.21MB | 44.21MB" type="" state=""><![CDATA[FirstPAGE Server]]></Name>
			</MonName>
			<MonName>
				<Name misc="256.87MB | 256.87MB | 256.87MB" type="" state=""><![CDATA[FPAM Gateway for ANS]]></Name>
			</MonName>
			<MonName>
				<Name misc="59.68MB | 59.68MB | 59.68MB" type="" state=""><![CDATA[Marquee Manager]]></Name>
			</MonName>
		</Item>
		<Item itemType="Space Allocated" stateEntries="" lastChecked="5/19/2022 12:29:28 PM">
			<MonName>
				<Name misc="72.42% Free" type="" state=""><![CDATA[C:]]></Name>
			</MonName>
			<MonName>
				<Name misc="86.43% Free" type="" state=""><![CDATA[D:]]></Name>
			</MonName>
			<MonName>
				<Name misc="93.80% Free" type="" state=""><![CDATA[E:]]></Name>
			</MonName>
		</Item>
		<Item itemType="Processor Usage" stateEntries="" lastChecked="5/19/2022 12:29:28 PM">
			<MonName>
				<Name misc="11.00% | 2.00% | 5.00%" type="" state=""><![CDATA[CPU (Total)]]></Name>
			</MonName>
		</Item>
	</Items>
</ItemsHeader>

```
I want to be able to grab the "misc", "state", CDATA values, LastChecked, all of that kind of stuff. So far I haven't had luck getting anything to return.

EDIT:
I made some progress on this after going back to some of the other discussions and trying a few things. I can get results now! But I haven't found an example on how to pull out the "misc" or "state" that are subtopics of Name.

![image](https://user-images.githubusercontent.com/869098/169371488-3f5a1eff-dea5-45f7-aad3-04ee331f7ee6.png)

**Comment by yesoreyeram** [comment, marked as answer]
Not sure what is your desired format. But, with UQL I am getting [this](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22https:%2F%2Fjsonplaceholder.typicode.com%2Fusers%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22data%22:%22%3CItemsHeader%3E%5Cn%5Ct%3CItems%3E%5Cn%5Ct%5Ct%3CItem%20itemType%3D%5C%22Runtime%20Parameters%5C%22%20stateEntries%3D%5C%22%5C%22%20lastChecked%3D%5C%225%2F19%2F2022%2012:29:28%20PM%5C%22%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%22ANS%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BAlertDestination%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%2214%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BArchiveExpiryDays%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%22LEFT%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BCheckNameAppended%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%22False%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BDebug%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%3C%2FItem%3E%5Cn%5Ct%5Ct%3CItem%20itemType%3D%5C%22Ping%20Results%5C%22%20stateEntries%3D%5C%22%5C%22%20lastChecked%3D%5C%225%2F19%2F2022%2012:29:28%20PM%5C%22%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%22OK%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BServer1%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%22OK%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BServer2%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%22OK%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BServer3%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%22OK%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BServer4%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%3C%2FItem%3E%5Cn%5Ct%5Ct%3CItem%20itemType%3D%5C%22Memory%20Usage%5C%22%20stateEntries%3D%5C%22%5C%22%20lastChecked%3D%5C%225%2F19%2F2022%2012:29:28%20PM%5C%22%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%2214.18MB%20%7C%2014.18MB%20%7C%2014.18MB%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BFirstPAGE%20Alarm%20Manager%20Server%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%2244.21MB%20%7C%2044.21MB%20%7C%2044.21MB%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BFirstPAGE%20Server%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%22256.87MB%20%7C%20256.87MB%20%7C%20256.87MB%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BFPAM%20Gateway%20for%20ANS%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%2259.68MB%20%7C%2059.68MB%20%7C%2059.68MB%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BMarquee%20Manager%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%3C%2FItem%3E%5Cn%5Ct%5Ct%3CItem%20itemType%3D%5C%22Space%20Allocated%5C%22%20stateEntries%3D%5C%22%5C%22%20lastChecked%3D%5C%225%2F19%2F2022%2012:29:28%20PM%5C%22%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%2272.42%25%20Free%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BC:%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%2286.43%25%20Free%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BD:%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%2293.80%25%20Free%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BE:%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%3C%2FItem%3E%5Cn%5Ct%5Ct%3CItem%20itemType%3D%5C%22Processor%20Usage%5C%22%20stateEntries%3D%5C%22%5C%22%20lastChecked%3D%5C%225%2F19%2F2022%2012:29:28%20PM%5C%22%3E%5Cn%5Ct%5Ct%5Ct%3CMonName%3E%5Cn%5Ct%5Ct%5Ct%5Ct%3CName%20misc%3D%5C%2211.00%25%20%7C%202.00%25%20%7C%205.00%25%5C%22%20type%3D%5C%22%5C%22%20state%3D%5C%22%5C%22%3E%3C!%5BCDATA%5BCPU%20(Total)%5D%5D%3E%3C%2FName%3E%5Cn%5Ct%5Ct%5Ct%3C%2FMonName%3E%5Cn%5Ct%5Ct%3C%2FItem%3E%5Cn%5Ct%3C%2FItems%3E%5Cn%3C%2FItemsHeader%3E%5Cn%22,%22uql%22:%22parse-xml%5Cn%7C%20scope%20%5C%22ItemsHeader.Items.Item%5C%22%5Cn%7C%20mv-expand%20%5C%22MonName%5C%22%5Cn%7C%20project%20%5C%22lastChecked%5C%22%3Dtodatetime(%5C%22@_lastChecked%5C%22),%20%5C%22type%5C%22%3D%5C%22@_itemType%5C%22,%20%5C%22Name%5C%22%3D%5C%22MonName.Name.%23text%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D)

```kusto
parse-xml
| scope "ItemsHeader.Items.Item"
| mv-expand "MonName"
| project "lastChecked"=todatetime("@_lastChecked"), "type"="@_itemType", "Name"="MonName.Name.#text"
```

<img width="1436" alt="image" src="https://user-images.githubusercontent.com/153843/169384070-846180ab-f2af-462c-b5de-1258f591ded8.png">

**Reply by gdiana** [reply]
Using XML can be tricky and a great problem is accessing variables above or below the chosen point in the hierarchy to try and create a flat structure.

UQL allows this to be overcome as may be seen above where variables lower down in the scope may be readily accessed but not those higher up, as far as I am aware?

Also, note that the variables may have an @ prepended or # or even renamed, so it is best to check the raw output by traversing the tree within GRAFANA as it may and will differ from any browser tools. 

I can assure you that this INFINITY datasource provides the most flexibility of all I have used, and I am restricted to XML because of the API I am required to use.

UQL also allows jsonata to be used which is more flexible and powerful and allows one to go up and down the hierarchy as shown below

**parse-xml
|jsonata "(ItemsHeader.Items.Item.MonName.Name.{'MISC' : `@_misc`, 'CDATA':`#text` , 'Item Type' : %.%.`@_itemType`, 'Last Checked' : %.%.`@_lastChecked` })"**

Just order the variables in the way you wish the columns to appear.

Here I started at the lowest level and then used the %.%.`variable` to access stuff further up, but it works both ways and produces the same result as shown below

![image](https://user-images.githubusercontent.com/13445152/169509192-1871aa9a-64f9-408c-bd6d-68ea38a8dae0.png)

I would suggest this datasource for what you are trying to do as it provides several elegant ways to "skin the cat"

**Reply by vulnox** [reply]
Thank you, both! That was extremely helpful and appreciate the additional background. Infinity is outstanding and I think what was provided is exactly what I need, I had not yet explored UQL and it seems like the right approach.

Thank you, again!

**Reply by gdiana** [reply]
I fiddled a bit with XML and the best I could do was the first Item. I am not sure how to select all the Items and have to use 0,1,2,3 

**MonName.[0]Name[0].$.misc        for     MISC
MonName.[0]Name[0]._                   for     CDATA**

![image](https://user-images.githubusercontent.com/13445152/169521819-c879ae6b-d312-49e5-b5f5-8e950b5d26c1.png)
